server.py

```diff
@@ -1,36 +1,376 @@
+        return super().__new__(
+            cls,
+            id=id,
+            name=name,
+            bio=bio,
+            available=available,
+            birth_year=birth_year,
+            gender=gender,
+            languages=languages,
+            need=need,
+            rights=rights,
+            housing=housing,
+        )
+
+    def to_json(self) -> CoachJSON:
+        return self._asdict()
 
-def check_request_for_authorization_status(request):
-    sanic.log.logger.info(f"Checking authorization: IP is {request.ip}")
-    match = (request.ip == "127.0.0.1")
-    sanic.log.logger.info(f"Check IP matches 127.0.0.1: {match}")
+    @classmethod
+    def from_json(cls, obj: CoachJSON) -> "Coach":
+        return cls(**obj)
 
-    return match
+    def copy(self) -> "Coach":
+        return self.from_json(self.to_json())
 
-def authorized(f):
-    @functools.wraps(f)
-    async def decorated_function(request, *args, **kwargs):
-        is_authorized = check_request_for_authorization_status(request)
 
-        if is_authorized:
-            # Run the handler method and return the response.
-            response = await f(request, *args, **kwargs)
-            return response
+class CoachDB:
+    """The in-memory coach database."""
+
+    def __init__(
+        self, file: Optional[os.PathLike] = None, *, auto_persist: bool = True
+    ):
+        """
+        :param file:
+            File to use as persistent storage of the DB. If not found it will be
+            created.
+        :param auto_persist:
+            Whether to automatically persist the DB to file on operations that
+            modify the DB. Ignored if file is not given.
+            Set this to False if manual batching is desired.
+        """
+        self._db: Dict[int, Coach] = dict()
+        self._file = file
+        self._auto_persist = auto_persist and file
+        self._next_id: int = 0
+        if file:
+            self._load_from_file(file)
+
+    def __iter__(self):
+        return iter(self._db.values())
+
+    def __contains__(self, item: Union[int, Coach]):
+        if item in self._db:
+            return True
+        elif isinstance(item, Coach):
+            return item.id in self._db
+        else:
+            return False
+
+    def add(self, coach: Coach) -> Coach:
+        """Add a new coach to the DB, setting the ID."""
+        assert self._next_id not in self._db
+        if coach.id is not None:
+            raise ValueError("Coach already has an ID assigned, not adding to DB")
+        db_coach = Coach(**{**coach.to_json(), "id": self._next_id})
+        self._next_id += 1
+        self._db[db_coach.id] = db_coach
+        logger.debug("Added coach with ID %d", coach.id)
+        if self._auto_persist:
+            self.persist(self._file)
+        return db_coach
+
+    def remove(self, coach_id: int):
+        """Remove a coach from the DB by ID."""
+        try:
+            self._db.pop(coach_id)
+        except KeyError:
```

Comment:
Unsure whether we should allow the KeyError to bubble up here - should we just catch it and log that the ID wasn't found? If the ID is supposed to be removed, and it isn't in the DB, then returning 'success' might be okay.

Reply:
Yeah APIs can go either way here - choice of idempotency or letting the user know (in a Pythonic way!) whether there was something to remove. I chose for this to behave like `list.remove()` and `set.remove()` (although admittedly the former raises a ValueError).

The expected usage if you don't care about whether the coach already existed is simply:
```
try:
    coach_db.remove(coach_id)
except KeyError:
    pass
```

Personally this choice feels right to me, but I don't feel that strongly. I would have to disagree with logging here though - you're effectively saying "callers can call this as much as they like on IDs that don't exist", which could mean it gets called a lot, and there's no need to log in all these valid uses of the API!

If your argument is then that callers should only call this when they *think* the ID exists and we should quietly handle the case they were wrong about that... then I'd disagree with that being good API design :)

Also note that if no KeyError were raised, we'd silently be accepting any nonsense for 'coach_id' (unless we checked the arg, which just feels untidy).

Reply:
Sounds good, I agree on second thoughts - let's keep it as it is!